src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/LocalCIDockerService.java

```diff
@@ -246,15 +255,103 @@ public void pullDockerImage(LocalCIBuildJobQueueItem buildJob, BuildLogsMap buil
+                if (dockerImageCleanupInfo.get(dockerImage).isBefore(ZonedDateTime.now().minusDays(imageExpiryDays))) {
+                    log.info("Deleting docker image {}", dockerImage);
+                    try {
+                        dockerClient.removeImageCmd(dockerImage).exec();
+                    }
+                    catch (NotFoundException e) {
+                        log.warn("Docker image {} not found during cleanup", dockerImage);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Checks for available disk space and triggers the cleanup of old Docker images if the available space falls below a specified threshold.
+     * The threshold is configurable via the 'artemis.continuous-integration.image-cleanup.disk-space-threshold-mb' property in the application settings.
+     *
+     * @implNote - We use the Docker root directory to check disk space availability. This is in case the Docker images are stored on a separate partition.
+     *           - We need to iterate over the map entries since don't remove the oldest image from the map.
+     */
+
+    @Scheduled(fixedRateString = "${artemis.continuous-integration.image-cleanup.disk-space-check-interval-ms:3600000}", initialDelayString = "${artemis.continuous-integration.image-cleanup.disk-space-check-interval-ms:3600000}")
+    public void checkUsableDiskSpaceThenCleanUp() {
+        if (imageCleanupEnabled) {
+            try {
+                // Get the Docker root directory to check disk space.
+                File file = new File(Objects.requireNonNullElse(dockerClient.infoCmd().exec().getDockerRootDir(), "/"));
+                long usableSpace = file.getUsableSpace();
+
+                // Get the threshold in bytes.
+                long bytesConversionFactor = 1024;
+                long threshold = imageCleanupDiskSpaceThresholdMb * bytesConversionFactor * bytesConversionFactor;
+
+                if (usableSpace >= threshold) {
+                    return;
+                }
+
+                // Get map of docker images and their last build dates
+                IMap<String, ZonedDateTime> dockerImageCleanupInfo = hazelcastInstance.getMap("dockerImageCleanupInfo");
+
+                // Get unused images
+                Set<String> unusedImages = getUnusedDockerImages();
+
+                // Get a sorted list of images by last build date
+                // We cast to ArrayList since we need the list to be mutable
+                List<Map.Entry<String, ZonedDateTime>> sortedImages = new java.util.ArrayList<>(
+                        dockerImageCleanupInfo.entrySet().stream().sorted(Map.Entry.comparingByValue()).toList());
+
+                if (sortedImages.isEmpty()) {
+                    return;
+                }
+
+                int maxDeleteAttempts = 5;
+                int maxTotalAttempts = sortedImages.size(); // We limit the total number of attempts to avoid infinite loops
```

Comment:
For me it is unintuitive that a "max..." value is modified during the loop

```suggestion
                int deleteAttempts = 5;
                int totalAttempts = sortedImages.size(); // We limit the total number of attempts to avoid infinite loops
```